### PR TITLE
feat: refactor: form validation logic

### PR DESCRIPTION
**Pull request by jlews**
## What's new?

Tiny MR for separating validation errors passed into the hook from validation errors dispatched client side.

(Also added a lint script to project root)

## Ticket number(s) in JIRA (if internal)

N/A

[board](https://rocketmakers.atlassian.net/jira/software/projects/HM/boards/172)

## Checklist

- [N/A] are your changes in Storybook?
- [N/A] are any breaking changes documented in `docs/migrating_from_oldstrong.md`?
- [N/A] does _everything_ have jsdoc?
- [N/A] is everything exported from index.ts?

### CODE REVIEW

**Comment by NathanRussellUK** [comment]
Code looks grand!

I wonder if client validation errors is the right word for it though, we spoke about generating validation errors in the body of the component that renders the form (in a memo), and passing those into the hook - that'd be on the client too

**Comment by patrickbenjamin1** [comment]
> Code looks grand!
> 
> I wonder if client validation errors is the right word for it though, we spoke about generating validation errors in the body of the component that renders the form (in a memo), and passing those into the hook - that'd be on the client too

Yeah I agree with that. You could theoretically do both client and server validation using either method (either the dispatcher on the formProp or the memo) so maybe naming them implying otherwise is something to avoid. It might still be preferable for some people to use the memoised option.

File: package.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "check-versions": "ncu && (cd module && ncu) && (cd playground && ncu)",
     "upgrade-all": "ncu -u && (cd module && ncu -u) && (cd playground && ncu -u) && npm i",
     "storybook": "(cd module && npm run storybook)",
-    "build-storybook": "(cd module && npm run build-storybook)"
+    "build-storybook": "(cd module && npm run build-storybook)",
+    "lint": "(cd module && npm run lint)"
```
Comment (patrickbenjamin1) [inline package.json:21]: ha i did that too in my branch

**Comment by jlews** [comment]
> > Code looks grand!
> > I wonder if client validation errors is the right word for it though, we spoke about generating validation errors in the body of the component that renders the form (in a memo), and passing those into the hook - that'd be on the client too
> 
> Yeah I agree with that. You could theoretically do both client and server validation using either method (either the dispatcher on the formProp or the memo) so maybe naming them in a way that implies otherwise is worth avoiding. It might still be preferable for some people to use the config argument option for client validation errors, say if they're using their own piece of state or something.
> 
> Like you could do this, using the dispatched for server validation errors, and I don't think it's actually that bad, especially if you're not using our API Hooks and you theoretically aren't using something that's storing your server errors in state:
> 
> ```tsx
> const submit = React.useCallback(async () => {
>   try {
>     const result = await doThing();
>   } catch(errors) {
>     const validationError = getValidationErrors(errors)
>     formProp('thing').dispatchValidationError(validationError)
>   }
> }, []);
> ```
> 
> And you could also still do this the old way, which is still fairly legit
> 
> ```tsx
> const [errors, setErrors] = useState()
> 
> const {} = Form.use({}, {validationErrors: errors})
> 
> const submit = React.useCallback(()=>{
>    if (data is bad){
>       setErrors([{ ..... }])
>    }
> })
> ```
> 
> I dunno what you'd call them to make the distinction instead is though...

Although I take the general point about the name, I don't see any urgent need to change it unless anyone can think of an alternative that will still make sense.

**Comment by NathanRussellUK** [comment]
Yep - not urgent/ shouldn't hold things up from my perspective.

As a bad stab at it(!), *hookValidationErrors* vs *formPropValidationErrors*
(or *controlledValidationErrors*, as they're set via methods rather than
directly)?

(might be great if those two examples went into docs somewhere if this
isn't covered already!!)


*Nathan Russell*
Web Developer @ Rocketmakers <http://www.rocketmakers.com/>

<http://www.rocketmakers.com/>

**Pull request merged**